File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.js
const input = require('fs').readFileSync('/dev/stdin').toString().split('\n');
const N = Number(input[0]);

const map = input.slice(1).map(line => line.split('').map(Number));
const visited = Array.from({ length: N }, () => Array(N).fill(false));
const result = [];

const dx = [0, 0, -1, 1];
const dy = [-1, 1, 0, 0];
function dfs(x, y) {
    let count = 1;
    visited[x][y] = true;

    for (let d = 0; d < 4; d++) {
        const nx = x + dx[d];
        const ny = y + dy[d];

        if (
            nx >= 0 && nx < N &&
            ny >= 0 && ny < N &&
            map[nx][ny] === 1 &&
            !visited[nx][ny]
        ) {
            count += dfs(nx, ny);
        }
    }

    return count;
}

for (let i = 0; i < N; i++) {
    for (let j = 0; j < N; j++) {
        if (map[i][j] === 1 && !visited[i][j]) {
            result.push(dfs(i, j));
        }
    }
}

result.sort((a,b) => a-b); // 오름차순 정렬
console.log(result.length);
console.log(result.join('\n'));

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
import sys
from collections import deque
input = sys.stdin.readline

N = int(input().strip())
graph = [list(map(int, input().strip())) for _ in range(N)]
visited = [[False] * N for _ in range(N)]

d_row = [0, 0, -1, 1]
d_col = [-1, 1, 0, 0]

def bfs(row, col):
    queue = deque([(row, col)])
    
    visited[row][col] = True
    count = 1
    
    while queue:
        r, c = queue.popleft()
        
        for i in range(4):
            nr, nc = r + d_row[i], c + d_col[i]
            
            if 0 <= nr < N and 0 <= nc < N and not visited[nr][nc] and graph[nr][nc] == 1:
                visited[nr][nc] = True
                queue.append((nr, nc))
                count += 1
                
    return count

result = []
for i in range(N):
    for j in range(N):
        if graph[i][j] == 1 and not visited[i][j]:
            result.append(bfs(i,j))
            
result.sort()
print(len(result))
print('\n'.join(map(str, result)))